stan/math/rev/functor/cvodes_integrator_adjoint.hpp

```diff
@@ -0,0 +1,829 @@
+
+    Eigen::Matrix<var, Eigen::Dynamic, 1> f_y_t_vars
+        = rhs(t, y_vec, local_args_tuple_);
+
+    check_size_match(solver_->function_name_, "dy_dt", f_y_t_vars.size(),
+                     "states", N_);
+
+    f_y_t_vars.adj() = -mu;
+
+    grad();
+
+    apply([&](auto&&... args) { accumulate_adjoints(mu_dot.data(), args...); },
+          local_args_tuple_);
+  }
+
+  /**
+   * Calculates the jacobian of the ODE RHS wrt to its states y at the
+   * given time-point t and state y.
+   */
+  inline void jacobian_states(double t, N_Vector y, SUNMatrix J) const {
+    Eigen::Map<Eigen::MatrixXd> Jfy(SM_DATA_D(J), N_, N_);
+    Eigen::Map<const Eigen::VectorXd> x(NV_DATA_S(y), N_);
+
+    nested_rev_autodiff nested;
+
+    const Eigen::Matrix<var, Eigen::Dynamic, 1> y_var(x);
+    Eigen::Matrix<var, Eigen::Dynamic, 1> fy_var
+        = rhs(t, y_var, value_of_args_tuple_);
+
+    check_size_match(solver_->function_name_, "dy_dt", fy_var.size(), "states",
+                     N_);
+
+    grad(fy_var.coeffRef(0).vi_);
+    Jfy.col(0) = y_var.adj();
+    for (int i = 1; i < fy_var.size(); ++i) {
+      nested.set_zero_all_adjoints();
+      grad(fy_var.coeffRef(i).vi_);
+      Jfy.col(i) = y_var.adj();
+    }
+    Jfy.transposeInPlace();
+  }
+
+  /*
+   * Calculate the Jacobian of the RHS of the adjoint ODE (see rhs_adj_sens
+   * below for citation for how this is done)
+   *
+   * @param[in] y State of system
+   * @param[in] t Time
+   * @param[out] J CVode structure where output is to be stored
+   */
+  inline void jacobian_adj(double t, N_Vector y, SUNMatrix J) const {
+    Eigen::Map<Eigen::MatrixXd> J_adj_y(SM_DATA_D(J), N_, N_);
+
+    // J_adj_y = -1 * transpose(J_y)
+    jacobian_states(t, y, J);
+
+    J_adj_y.transposeInPlace();
+    J_adj_y.array() *= -1.0;
+  }
+
+  /**
+   * Overloads which setup the states returned from the forward solve. In case
+   * the return type is a double only, then no autodiff is needed. In case of
+   * autodiff then non-chaining varis are setup accordingly.
+   */
+  void store_state(std::size_t n, const Eigen::VectorXd& state,
+                   Eigen::Matrix<var, Eigen::Dynamic, 1>& state_return) {
+    solver_->y_[n] = state;
+    state_return.resize(N_);
+    for (size_t i = 0; i < N_; i++) {
+      non_chaining_varis_[N_ * n + i] = new vari(state.coeff(i), false);
+      state_return.coeffRef(i) = var(non_chaining_varis_[N_ * n + i]);
+    }
+  }
+
+  void store_state(std::size_t n, const Eigen::VectorXd& state,
+                   Eigen::Matrix<double, Eigen::Dynamic, 1>& state_return) {
+    solver_->y_[n] = state;
+    state_return = state;
+  }
+
+  /**
+   * Since the CVODES solver manages memory with malloc calls, these resources
+   * must be freed using a destructor call (which is not being called for the
+   * vari class).
+   */
+  struct cvodes_solver : public chainable_alloc {
+    const size_t N_;
+    std::vector<Eigen::VectorXd> y_;
+    const std::string function_name_str_;
```

Comment:
idt this is used (and there's no reason to have const members here)

Reply:
Not sure I understand. Why would I not want to make things const which never change? And: The function name sting needs to be stored in a string, I think.

Reply:
Yeah it's fine, usually I'm anti-const members because const members cannot be moved from (you need to copy) but we never move/copy the actual object here (only the pointer) so having const members is fine